pkg/hoist/artifact_cache.go

```
@@ -0,0 +1,117 @@
+// whether the artifact is still in use by a particular Launchable (say, as the Last link). However,
+// as a result of this strategy, pruning the cache dir will not be sufficient to free system disk
+// resources as other links to the same tars will also need to be removed.
+type ArtifactCacher struct {
+	cacheDir      string
+	nestedFetcher Fetcher
+	Logger        *logging.Logger
+}
+
+// Complies with the Fetcher interface
+func (a *ArtifactCacher) Fetch(fromURI, toURI string) error {
+	base := filepath.Base(toURI)
+	if base == "." || base == string(filepath.Separator) {
+		return util.Errorf("%s is not a valid toURI", toURI)
+	}
+	cachePath := filepath.Join(a.cacheDir, base)
+	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
+		a.nestedFetcher(fromURI, cachePath)
+	} else if err != nil {
+		return util.Errorf("Error stating the cache path: %s", err)
+	} else {
+		err = a.touch(cachePath)
+		if err != nil {
+			a.Logger.WithField("err", err).Warnln("Error when updating mtime on cached artifact")
+		}
+	}
+	// hardlink so we can safely prune without breaking Last/Current links
+	linkErr := os.Link(cachePath, toURI)
+	if linkErr != nil {
+		// we may be on different mounts, just copy
+		err := uri.URICopy(cachePath, toURI)
+		if err != nil {
+			return util.Errorf("Could not link or copy file from %s to %s: %s - %s", cachePath, toURI, err, linkErr)
+		}
+	}
+	return nil
+}
+
+type byMtime []os.FileInfo
+
+func (b byMtime) Len() int           { return len(b) }
+func (b byMtime) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+func (b byMtime) Less(i, j int) bool { return b[i].ModTime().Before(b[j].ModTime()) }
+
+// Prune will remove all but `keep` # of artifacts
+// from the cache dir. Prune will favor more recently modified artifacts
+func (a *ArtifactCacher) Prune(keep int) {
+	infos, err := ioutil.ReadDir(a.cacheDir)
+	if err != nil {
+		err = util.Errorf("cache dir unlistable: %s", err)
+		a.Logger.WithField("err", err).Errorln("Pruning failed")
```

why not just return these errors from `Prune` and delegate logging to the caller?